File: Solutions/Python/Problem 016.py
# Project Euler Problem 16 - Power Digit Sum
# https://projecteuler.net/problem=16
# Answer = 1366

def question():
    """
    Print the problem statement.
    """
    print("""Find the sum of the digits of the result of 2^1000.""")

def solve(base, exponent):
    """
    Calculate the result of base^exponent and return the sum of its digits.
    Time Complexity: O(log(exponent))

    Args:
        base (int): The base number.
        exponent (int): The exponent.

    Returns:
        int: The sum of the digits of base^exponent.
    """
    num_list = map(int, str(pow(base, exponent)))
    return sum(num_list)

def main():
    question()
    result = solve(2, 1000)
    print(f"The answer is {result}")

if __name__ == "__main__":
    main()

File: Solutions/Rust/problem_016.rs
// Problem 16 - Power digit sum
// https://projecteuler.net/problem=16
// Answer = 1366

fn question() {
    println!("2^15 = 32768 and the sum of its digits is 3 + 2 + 7 + 6 + 8 = 26.");
    println!("What is the sum of the digits of the number 2^1000?");
}

/// Calculates the sum of the individual digits of 2 raised to a given power.
///
/// # Arguments
/// 
/// * `power` - The exponent to which the base 2 is raised.
///
/// # Returns
///
/// The scalar sum of all decimal digits in the resulting expansion.
fn solve(base:u32, power: u32) -> u32 {
    let capacity = (f64::from(power) * f64::from(base).log10()).ceil() as usize + 1;
    let mut power_answer: Vec<u8> = Vec::with_capacity(capacity);
    power_answer.push(1);
    for _exponent in 0..power{
        let mut carry = 0;
        for digit in power_answer.iter_mut(){
            let product = (base * (*digit as u32)) + carry;
            *digit = (product % 10) as u8;
            carry = product / 10;
        }
        while carry > 0 {
            power_answer.push((carry % 10) as u8);
            carry /= 10;
        }
    }
    power_answer.iter().map(|&x| x as u32).sum()
}

fn main() {
    question();
    let result = solve(2, 1000);
    println!("Answer = {}", result);
}


// TODO: Implement macro-stepping optimization by finding the maximum 
// power 'k' such that (base^k) safely fits within a u128 accumulator.